rcl_action/test/rcl_action/test_graph.cpp

```diff
@@ -0,0 +1,548 @@
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    this->old_node = rcl_get_zero_initialized_node();
+    rcl_node_options_t node_options = rcl_node_get_default_options();
+    ret = rcl_node_init(
+      &this->old_node, this->test_graph_old_node_name, "", &this->old_context, &node_options);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    ret = rcl_shutdown(&this->old_context);  // after this, the old_node should be invalid
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+    this->context = rcl_get_zero_initialized_context();
+    ret = rcl_init(0, nullptr, &init_options, &this->context);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    this->node = rcl_get_zero_initialized_node();
+    ret = rcl_node_init(&this->node, test_graph_node_name, "", &this->context, &node_options);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  }
+
+  void TearDown()
+  {
+    rcl_ret_t ret;
+    ret = rcl_node_fini(&this->old_node);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+    ret = rcl_node_fini(&this->node);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+    ret = rcl_shutdown(&this->context);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    ret = rcl_context_fini(&this->context);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    ret = rcl_context_fini(&this->old_context);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  }
+};
+
+TEST_F(
+  CLASSNAME(TestActionGraphFixture, RMW_IMPLEMENTATION),
+  test_action_get_client_names_and_types_by_node)
+{
+  rcl_ret_t ret;
+  rcl_node_t zero_node = rcl_get_zero_initialized_node();
+  rcl_names_and_types_t nat = rcl_get_zero_initialized_names_and_types();
+
+  // Invalid node
+  ret = rcl_action_get_client_names_and_types_by_node(
+    nullptr, &this->allocator, this->test_graph_node_name, "", &nat);
+  EXPECT_EQ(RCL_RET_NODE_INVALID, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  ret = rcl_action_get_client_names_and_types_by_node(
+    &zero_node, &this->allocator, this->test_graph_node_name, "", &nat);
+  EXPECT_EQ(RCL_RET_NODE_INVALID, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  ret = rcl_action_get_client_names_and_types_by_node(
+    &this->old_node, &this->allocator, this->test_graph_node_name, "", &nat);
+  EXPECT_EQ(RCL_RET_NODE_INVALID, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  // Invalid allocator
+  ret = rcl_action_get_client_names_and_types_by_node(
+    &this->node, nullptr, this->test_graph_node_name, "", &nat);
+  EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  // Invalid node name
+  ret = rcl_action_get_client_names_and_types_by_node(
+    &this->node, &this->allocator, "_test_this_Isnot_a_valid_name", "", nullptr);
+  EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret) << rcl_get_error_string().str;
```

Comment:
After passing `&nat` here, I released that it actually returns `RCL_RET_ERROR`. This because of the underlying call to `rcl_get_subscriber_names_and_types_by_node`, which has the same behavior.